open_instruct/tool_utils/tool_vllm.py

```diff
@@ -0,0 +1,304 @@
+"""
+python open_instruct/tool_utils/tool_vllm.py
+"""
+from collections import defaultdict
+from dataclasses import dataclass
+import re
+import requests
+import traceback
+from vllm import LLM, SamplingParams, RequestOutput, PoolingRequestOutput, TokensPrompt
+from tqdm import tqdm
+from typing import Union
+from rich.console import Console
+from concurrent.futures import ThreadPoolExecutor  # add import for async execution
+
+@dataclass
+class ToolOutput:
+    output: str
+    called: bool
+    success: bool
+
+class Tool:
+    def __init__(self, start_str: str, end_str: str):
+        self.start_str = start_str
+        self.end_str = end_str
+
+    def __call__(self, prompt: str) -> ToolOutput:
+        raise NotImplementedError("Subclasses must implement this method")
+
+class PythonCodeTool(Tool):
+    """@vwxyzjn: I recommend using something like a FastAPI for this kind of stuff; 1) you 
+    won't accidentally block the main vLLM process and 2) way easier to parallelize via load balancing."""
+    def __call__(self, prompt: str) -> ToolOutput:
+        # Find Python code blocks using regex
+        re_str = r'<tool>\s*(.*?)\s*</tool>'
```

Comment:
Wondering if we can parametrize some of the special tokens to ascribe w/ the current template: https://github.com/allenai/open-instruct/blob/toolu/open_instruct/dataset_transformation.py#L251

there might be a chance the template might change though!